<launch>
    <!-- Task and logging arguments -->
    <arg name="task_type"/>
    <arg name="logging_enabled"                 default="true"/>
    <arg name="test_id"                         default="default"/>

    <!-- Simulator arguments -->
    <arg name="launch_simulator"                default="true"/>
    <arg name="feedback_covariance"             default="0.0"/>
    <arg name="start_bullet_viewer"             default="true"/>
    <arg name="screenshots_enabled"             default="false"/>

    <!-- Planner model arguments -->
    <arg name="multi_model"                     default="false"/>
    <arg name="deformability_override"          default="false"/>
    <arg name="translational_deformability"     if="$(arg deformability_override)"/>
    <arg name="rotational_deformability"        if="$(arg deformability_override)"/>
    <arg name="use_adaptive_model"              default="false"/>
    <arg name="adaptive_model_learning_rate"    default="0.000001"/>
    <arg name="calculate_regret"                default="false"/>

    <!-- Planner internal arguments -->
    <arg name="launch_planner"                          default="true"/>
    <arg name="planning_horizon"                        default="1"/>
    <arg name="optimization_enabled"                    default="true"/>
    <arg name="kalman_parameters_override"              default="false"/>
    <arg name="process_noise_factor"                    if="$(arg kalman_parameters_override)"/>
    <arg name="observation_noise_factor"                if="$(arg kalman_parameters_override)"/>
    <arg name="bandit_algorithm"                        default="KFMANB"/>
    <arg name="use_random_seed"                         default="false"/>
    <arg name="static_seed_override"                    default="false"/>
    <arg name="static_seed"                             if="$(arg static_seed_override)"/>
    <arg name="correlation_strength_factor_override"    default="false"/>
    <arg name="correlation_strength_factor"             if="$(arg static_seed_override)"/>
    <arg name="max_correlation_strength_factor"         default="0.9"/>

    <!-- Setup task parameters -->
    <include file="$(find smmap)/launch/$(arg task_type)_params.launch"/>
    <param name="logging_enabled"                       value="$(arg logging_enabled)"                                      type="bool"/>
    <param name="log_folder"                            value="$(find smmap)/../logs/$(arg task_type)/$(arg test_id)/"      type="string"/>

    <!-- Launch the simulator -->
    <group ns="deform_simulator_node">
        <param name="feedback_covariance"               value="$(arg feedback_covariance)"          type="double"/>
        <param name="start_bullet_viewer"               value="$(arg start_bullet_viewer)"          type="bool"/>
        <param name="screenshots_enabled"               value="$(arg screenshots_enabled)"          type="bool"/>
    </group>

    <env name="OSG_FILE_PATH" value="$(find OpenSceneGraph)/data"/>
    <node name="deform_simulator_node" pkg="deform_simulator" type="custom_scene_node" required="true" if="$(arg launch_simulator)"/>

    <!-- Start the planner -->
    <group ns="smmap_planner_node">
        <param name="planning_horizon"                  value="$(arg planning_horizon)"                 type="int"/>
        <param name="optimization_enabled"              value="$(arg optimization_enabled)"             type="bool"/>
        <param name="use_multi_model"                   value="$(arg multi_model)"                      type="bool"/>
        <param name="translational_deformability"       value="$(arg translational_deformability)"      type="double"   if="$(arg deformability_override)"/>
        <param name="rotational_deformability"          value="$(arg rotational_deformability)"         type="double"   if="$(arg deformability_override)"/>
        <param name="use_adaptive_model"                value="$(arg use_adaptive_model)"               type="bool"/>
        <param name="adaptive_model_learning_rate"      value="$(arg adaptive_model_learning_rate)"     type="double"/>
        <param name="process_noise_factor"              value="$(arg process_noise_factor)"             type="double"   if="$(arg kalman_parameters_override)"/>
        <param name="observation_noise_factor"          value="$(arg observation_noise_factor)"         type="double"   if="$(arg kalman_parameters_override)"/>
        <param name="calculate_regret"                  value="$(arg calculate_regret)"                 type="bool"/>
        <param name="use_random_seed"                   value="$(arg use_random_seed)"                  type="bool"/>
        <param name="static_seed"                       value="$(arg static_seed)"                      type="string"   if="$(arg static_seed_override)"/>
        <param name="correlation_strength_factor"       value="$(arg correlation_strength_factor)"      type="double"   if="$(arg correlation_strength_factor_override)"/>
        <param name="max_correlation_strength_factor"   value="$(arg max_correlation_strength_factor)"  type="double"/>
    </group>

    <node name="smmap_planner_node" pkg="smmap" type="smmap_test_node_$(arg bandit_algorithm)" required="true" if="$(arg launch_planner)"/>
</launch>
